File: workspace/projects/heartwood-mcp/src/tools/jt/tasks.ts
/**
 * JT Task tools — 7 tools
 */

import type { PaveClient, ToolResult } from "../../pave/index.js";
import { TASK_FIELDS, TEMPLATE_FIELDS } from "../../pave/index.js";
import type { ToolDef } from "../registry.js";
import { buildFilter, pickDefined, requireString, PAGINATION_PROPS, getPagination } from "./helpers.js";

export function taskTools(pave: PaveClient): ToolDef[] {
  return [
    // ── jt_create_task ─────────────────────────────────────────────────
    {
      name: "jt_create_task",
      description: "Create a task on a target (job, document, etc.).",
      inputSchema: {
        type: "object" as const,
        properties: {
          name: { type: "string", description: "Task name" },
          targetType: { type: "string", description: "Target entity type (e.g., 'job')" },
          targetId: { type: "string", description: "Target entity ID (optional)" },
          description: { type: "string", description: "Task description (optional)" },
          assignees: {
            type: "array", items: { type: "string" },
            description: "User IDs to assign (optional)",
          },
          startDate: { type: "string", description: "Start date YYYY-MM-DD (optional)" },
          endDate: { type: "string", description: "End date YYYY-MM-DD (optional)" },
          isToDo: { type: "boolean", description: "Is this a to-do item (optional)" },
          isGroup: { type: "boolean", description: "Is this a task group (optional)" },
          progress: { type: "number", description: "Progress 0-100 (optional)" },
        },
        required: ["name", "targetType"],
      },
      handler: async (params: Record<string, unknown>): Promise<ToolResult> => {
        const data: Record<string, unknown> = {
          name: params.name,
          targetType: params.targetType,
          ...pickDefined(params, [
            "targetId", "description", "assignees", "startDate",
            "endDate", "isToDo", "isGroup", "progress",
          ]),
        };
        return pave.create("task", data, TASK_FIELDS);
      },
    },

    // ── jt_update_task_progress ────────────────────────────────────────
    {
      name: "jt_update_task_progress",
      description: "Update a task's progress, name, description, or dates.",
      inputSchema: {
        type: "object" as const,
        properties: {
          taskId: { type: "string", description: "Task ID" },
          progress: { type: "number", description: "Progress 0-100 (optional)" },
          name: { type: "string", description: "New name (optional)" },
          description: { type: "string", description: "New description (optional)" },
          startDate: { type: "string", description: "Start date YYYY-MM-DD (optional)" },
          endDate: { type: "string", description: "End date YYYY-MM-DD (optional)" },
        },
        required: ["taskId"],
      },
      handler: async (params: Record<string, unknown>): Promise<ToolResult> => {
        const id = requireString(params, "taskId");
        if ("error" in id) return id.error;
        const data = pickDefined(params, ["progress", "name", "description", "startDate", "endDate"]);
        return pave.update("task", id.value, data, TASK_FIELDS);
      },
    },

    // ── jt_get_tasks ───────────────────────────────────────────────────
    {
      name: "jt_get_tasks",
      description: "Get tasks, optionally filtered by job, status, or assignee.",
      inputSchema: {
        type: "object" as const,
        properties: {
          jobId: { type: "string", description: "Filter by job ID (optional)" },
          status: { type: "string", description: "Filter by status (optional)" },
          assigneeUserId: { type: "string", description: "Filter by assignee user ID (optional)" },
          ...PAGINATION_PROPS,
        },
        required: [],
      },
      handler: async (params: Record<string, unknown>): Promise<ToolResult> => {
        return pave.query({
          entity: "task",
          fields: TASK_FIELDS,
          filter: buildFilter(params, [
            { param: "jobId", field: "jobId" },
            { param: "status", field: "status" },
            { param: "assigneeUserId", field: "assigneeUserId" },
          ]),
          ...getPagination(params),
        });
      },
    },

    // ── jt_get_task_details ────────────────────────────────────────────
    {
      name: "jt_get_task_details",
      description: "Get full task details including assignees and dependencies.",
      inputSchema: {
        type: "object" as const,
        properties: {
          taskId: { type: "string", description: "Task ID" },
        },
        required: ["taskId"],
      },
      handler: async (params: Record<string, unknown>): Promise<ToolResult> => {
        const id = requireString(params, "taskId");
        if ("error" in id) return id.error;
        return pave.read("task", id.value, [
          ...TASK_FIELDS,
          { field: "dependencies", fields: [{ field: "id" }, { field: "name" }] },
        ]);
      },
    },

    // ── jt_get_schedule_templates ──────────────────────────────────────
    {
      name: "jt_get_schedule_templates",
      description: "Get all schedule templates.",
      inputSchema: {
        type: "object" as const,
        properties: {},
        required: [],
      },
      handler: async (): Promise<ToolResult> => {
        return pave.query({
          entity: "scheduleTemplate",
          fields: TEMPLATE_FIELDS,
        });
      },
    },

    // ── jt_get_todo_templates ──────────────────────────────────────────
    {
      name: "jt_get_todo_templates",
      description: "Get all to-do templates.",
      inputSchema: {
        type: "object" as const,
        properties: {},
        required: [],
      },
      handler: async (): Promise<ToolResult> => {
        return pave.query({
          entity: "todoTemplate",
          fields: TEMPLATE_FIELDS,
        });
      },
    },

    // ── jt_get_task_template_details ───────────────────────────────────
    {
      name: "jt_get_task_template_details",
      description: "Get the tasks inside a specific task template.",
      inputSchema: {
        type: "object" as const,
        properties: {
          taskTemplateId: { type: "string", description: "Task template ID" },
        },
        required: ["taskTemplateId"],
      },
      handler: async (params: Record<string, unknown>): Promise<ToolResult> => {
        const id = requireString(params, "taskTemplateId");
        if ("error" in id) return id.error;
        return pave.read("taskTemplate", id.value, [
          ...TEMPLATE_FIELDS,
          { field: "tasks", fields: TASK_FIELDS },
        ]);
      },
    },
  ];
}

File: heartwood-mcp/tools/tasks.py
"""Task tools — create, update, and query JobTread tasks."""

from __future__ import annotations

from app import mcp
from constants import ORG_ID
from pave import call_pave, flatten_nodes


TASK_FIELDS = {
    "nodes": {
        "id": {},
        "name": {},
        "status": {},
        "progress": {},
        "jobId": {},
        "assigneeId": {},
        "startDate": {},
        "endDate": {},
        "description": {},
    }
}


@mcp.tool()
async def jt_create_task(
    job_id: str,
    name: str,
    assignee_id: str | None = None,
    start_date: str | None = None,
    end_date: str | None = None,
    description: str | None = None,
) -> dict:
    """Create a task on a job.

    Args:
        job_id: The job ID.
        name: Task name.
        assignee_id: User ID to assign the task to.
        start_date: Start date (YYYY-MM-DD).
        end_date: End date (YYYY-MM-DD).
        description: Task description.
    """
    params: dict = {
        "organizationId": ORG_ID,
        "jobId": job_id,
        "name": name,
    }
    if assignee_id:
        params["assigneeId"] = assignee_id
    if start_date:
        params["startDate"] = start_date
    if end_date:
        params["endDate"] = end_date
    if description:
        params["description"] = description

    return await call_pave(
        "createTask",
        params=params,
        fields={"id": {}, "name": {}, "status": {}},
    )


@mcp.tool()
async def jt_update_task_progress(
    task_id: str,
    progress: int,
    status: str | None = None,
) -> dict:
    """Update a task's progress percentage and optionally its status.

    Args:
        task_id: The task ID.
        progress: Progress percentage (0-100).
        status: New status (e.g. "pending", "in_progress", "completed").
    """
    params: dict = {"id": task_id, "progress": progress}
    if status is not None:
        params["status"] = status

    return await call_pave(
        "updateTask",
        params=params,
        fields={"id": {}, "name": {}, "status": {}, "progress": {}},
    )


@mcp.tool()
async def jt_get_tasks(
    job_id: str | None = None,
    status: str | None = None,
    assignee_id: str | None = None,
    first: int = 100,
) -> list[dict]:
    """List tasks, optionally filtered by job, status, or assignee.

    Args:
        job_id: Filter by job.
        status: Filter by status.
        assignee_id: Filter by assigned user.
        first: Max results.
    """
    filter_obj: dict = {}
    if job_id:
        filter_obj["jobId"] = {"eq": job_id}
    if status:
        filter_obj["status"] = {"eq": status}
    if assignee_id:
        filter_obj["assigneeId"] = {"eq": assignee_id}

    params: dict = {"first": first}
    if filter_obj:
        params["filter"] = filter_obj

    result = await call_pave("tasks", params=params, fields=TASK_FIELDS)
    return flatten_nodes(result)


@mcp.tool()
async def jt_get_task_details(task_id: str) -> dict:
    """Get full details for a specific task.

    Args:
        task_id: The task ID.
    """
    result = await call_pave(
        "tasks",
        params={"filter": {"id": {"eq": task_id}}, "first": 1},
        fields={
            "nodes": {
                "id": {},
                "name": {},
                "status": {},
                "progress": {},
                "jobId": {},
                "job": {"id": {}, "name": {}, "number": {}},
                "assigneeId": {},
                "assignee": {"id": {}, "firstName": {}, "lastName": {}},
                "startDate": {},
                "endDate": {},
                "description": {},
                "createdAt": {},
                "updatedAt": {},
            }
        },
    )
    nodes = flatten_nodes(result)
    return nodes[0] if nodes else {}


@mcp.tool()
async def jt_get_task_templates(first: int = 50) -> list[dict]:
    """Get available task templates for the organization.

    Args:
        first: Max results.
    """
    result = await call_pave(
        "taskTemplates",
        params={"filter": {"organizationId": {"eq": ORG_ID}}, "first": first},
        fields={
            "nodes": {
                "id": {},
                "name": {},
                "description": {},
            }
        },
    )
    return flatten_nodes(result)
